aa/pb.py

```diff
@@ -295,13 +296,79 @@ class PbFileFetcher(fetcher.Fetcher):
         self._root = root
 
-    def _get_pb_file(self, pv, year):
-        # Split PV on either dash or colon
+    def _get_all_pb_files_of_pv(self, pv):
+        """Returns a list of all .pb files of the given PV which are found under the
+        root directory. Assuming the storage is split into STS/MTS/LTS."""
+        # Strip any STS/MTS/LTS from the root first to ensure backwards compatibility.
+        if os.path.basename(os.path.normpath(self._root)) in ["STS", "MTS", "LTS"]:
+            self._root = os.path.dirname(os.path.normpath(self._root))
+        pv_files = []
+        # Split PV on either dash or colon.
         parts = re.split("[-:]", pv)
         suffix = parts.pop()
-        directory = os.path.join(self._root, os.path.sep.join(parts))
-        filename = "{}:{}.pb".format(suffix, year)
-        return os.path.join(directory, filename)
+        # Go through long/mid/short term storages and collect all files for this PV.
+        for s in ["LTS", "MTS", "STS"]:
+            directory = os.path.join(self._root, s, os.path.sep.join(parts))
+            for f in sorted(glob.glob(os.path.join(directory, f"{suffix}*pb"))):
+                pv_files.append(f)
+        return pv_files
+
+    def _create_datetime_for_pb_file(self, filepath):
+        """Each .pb files of the archiver appliance ends with a date information
+        corresponding to the stored data. This function returns a datetime.datetime
+        object matching the date information of the given .pb file."""
+        filename = os.path.basename(filepath)
+        # The filename can contain just the year or stepwise more information up to
+        # year, month, day, hour and minutes.
+        dates = (
+            re.search(r"\d{4}(_\d{2})?(_\d{2})?(_\d{2})?(_\d{2})?", filename)
+            .group(0)
+            .split("_")
+        )
+        dates = [int(date) for date in dates]
+        # Make sure to give a least 3 arguments to datetime.datetime.
+        while len(dates) < 3:
+            dates.append(1)
+        return datetime.datetime(*dates, tzinfo=pytz.utc)
+
+    def _get_pb_files(self, pv, start, end=None):
+        """Goes through LTS/MTS/STS directories of given PV and returns all .pb files
+        containing data of the specified time window. The granularity setup of the
+        archiver for LTS/MTS/STS can deviate from the standard one."""
+        if end is None:
+            end = datetime.datetime.now().astimezone(pytz.utc)
+            log.info("No end time given, assuming now() in UTC.")
+        if start > end:
+            start, end = end, start
+            log.warning("End date was before start date. Swapped them.")
+        pv_files = self._get_all_pb_files_of_pv(pv)
+        pv_files_datetime = [self._create_datetime_for_pb_file(f) for f in pv_files]
+        # Find the files we need between start and end.
+        start_index, end_index = None, None
+        for i, file_date in enumerate(pv_files_datetime):
```

Comment:
You may be able to make this a bit simpler if you do something like:
```
matching_paths = []
for file_path in pv_files:
    file_date = self._create_datetime_for_pb_file(file_path)
    if file_date >= start and file_date <= end:
        matching_paths.append(file_path)
```

Reply:
A perfect example of "thinking too complex" :sweat_smile: thanks for this simplification.
I just added the case where also start > file_date of the last file. Depending of the time delta and the set granularity there might be already data from the given time window in the file.